src/ScopeProvider/ScopeProvider.tsx

```diff
@@ -0,0 +1,111 @@
+import { Provider, useStore } from 'jotai/react';
+import {
+  createContext,
+  useContext,
+  useState,
+  type PropsWithChildren,
+} from 'react';
+import { createScope, type Scope } from './scope';
+import type { AnyAtom, Store } from './types';
+
+const ScopeContext = createContext<{
+  scope: Scope | undefined;
+  baseStore: Store | undefined;
+}>({ scope: undefined, baseStore: undefined });
+
+export const ScopeProvider = ({
+  atoms,
+  children,
+  debugName,
+}: PropsWithChildren<{ atoms: Iterable<AnyAtom>; debugName?: string }>) => {
+  const parentStore: Store = useStore();
+  let { scope: parentScope, baseStore = parentStore } =
+    useContext(ScopeContext);
+  // if this scope is the first descendant scope under Provider then we don't want to inherit parentScope
+  // https://github.com/jotaijs/jotai-scope/pull/33#discussion_r1604268003
+  if (isTopLevelScope(parentStore)) {
```

Comment:
I wonder if we can distinguish top-level with ScopeContext without using isPatchedStore symbol.

Reply:
Hm, I thought the top level is `undefined`, but maybe it's the first level.
In that case, I'd add `isTop` of `level` property in the context value.
Anyway, I don't follow all details, and it's not necessary for this PR.